File: checkers_game.py
import numpy as np

class CheckersGame:
    """Custom Checkers game environment"""
    
    # Board piece representations
    EMPTY = 0
    RED = 1
    BLACK = 2
    RED_KING = 3
    BLACK_KING = 4

    def __init__(self):
        self.board = None
        self.current_player = None
        self.reset()

    def reset(self):
        """Initialize the game board"""
        self.board = np.zeros((8, 8), dtype=int)
        self.current_player = self.RED
        
        # Place black pieces (top of board)
        for row in range(3):
            for col in range(8):
                if (row + col) % 2 == 1:
                    self.board[row, col] = self.BLACK

        # Place red pieces (bottom of board)
        for row in range(5, 8):
            for col in range(8):
                if (row + col) % 2 == 1:
                    self.board[row, col] = self.RED

        return self.board.copy()

    def get_valid_moves(self, player):
        """Get all valid moves for the given player"""
        moves = []
        # First check for any available jumps (mandatory in checkers)
        for row in range(8):
            for col in range(8):
                if self.board[row, col] in [player, player + 2]:  # Regular or king piece
                    jumps = self._get_valid_jumps(row, col)
                    if jumps:
                        moves.extend(jumps)
        
        # If no jumps available, get regular moves
        if not moves:
            for row in range(8):
                for col in range(8):
                    if self.board[row, col] in [player, player + 2]:
                        moves.extend(self._get_valid_regular_moves(row, col))
        
        return moves

    def _get_valid_jumps(self, row, col):
        """Get all valid jump moves for a piece"""
        jumps = []
        piece = self.board[row, col]
        
        # Define possible jump directions based on piece type
        directions = []
        if piece == self.RED:
            directions.extend([(-2, -2), (-2, 2)])  # Up-left and up-right jumps only
        elif piece == self.BLACK:
            directions.extend([(2, -2), (2, 2)])    # Down-left and down-right jumps only
        elif piece in [self.RED_KING, self.BLACK_KING]:
            # Kings can jump in all four directions
            directions.extend([
                (-2, -2),  # Up-left
                (-2, 2),   # Up-right
                (2, -2),   # Down-left
                (2, 2)     # Down-right
            ])

        for dy, dx in directions:
            new_row, new_col = row + dy, col + dx
            mid_row, mid_col = row + dy//2, col + dx//2
            
            if self._is_valid_position(new_row, new_col):
                # Check if jump is valid
                if (self.board[new_row, new_col] == self.EMPTY and
                    self._is_opponent_piece(mid_row, mid_col, piece)):
                    jumps.append(((row, col), (new_row, new_col)))

        return jumps

    def _get_valid_regular_moves(self, row, col):
        """Get all valid regular moves for a piece"""
        moves = []
        piece = self.board[row, col]
        
        # Define possible move directions based on piece type
        directions = []
        if piece == self.RED:
            directions.extend([(-1, -1), (-1, 1)])  # Up-left and up-right only
        elif piece == self.BLACK:
            directions.extend([(1, -1), (1, 1)])    # Down-left and down-right only
        elif piece in [self.RED_KING, self.BLACK_KING]:
            # Kings can move in all four directions
            directions.extend([
                (-1, -1),  # Up-left
                (-1, 1),   # Up-right
                (1, -1),   # Down-left
                (1, 1)     # Down-right
            ])

        for dy, dx in directions:
            new_row, new_col = row + dy, col + dx
            if (self._is_valid_position(new_row, new_col) and 
                self.board[new_row, new_col] == self.EMPTY):
                moves.append(((row, col), (new_row, new_col)))

        return moves

    def _is_valid_position(self, row, col):
        """Check if position is within board boundaries"""
        return 0 <= row < 8 and 0 <= col < 8

    def _is_opponent_piece(self, row, col, piece):
        """Check if position contains an opponent's piece"""
        if not self._is_valid_position(row, col):
            return False
        target = self.board[row, col]
        return ((piece in [self.RED, self.RED_KING] and target in [self.BLACK, self.BLACK_KING]) or
                (piece in [self.BLACK, self.BLACK_KING] and target in [self.RED, self.RED_KING]))

    def make_move(self, move):
        """Execute a move on the board"""
        (start_row, start_col), (end_row, end_col) = move
        piece = self.board[start_row, start_col]
        
        # Move the piece
        self.board[start_row, start_col] = self.EMPTY
        self.board[end_row, end_col] = piece
        
        # Handle jumps (capture)
        if abs(end_row - start_row) == 2:
            mid_row = (start_row + end_row) // 2
            mid_col = (start_col + end_col) // 2
            self.board[mid_row, mid_col] = self.EMPTY
        
        # Handle king promotion
        if piece == self.RED and end_row == 0:
            self.board[end_row, end_col] = self.RED_KING
        elif piece == self.BLACK and end_row == 7:
            self.board[end_row, end_col] = self.BLACK_KING

    def is_game_over(self):
        """Check if the game is over (no valid moves for current player)"""
        return len(self.get_valid_moves(self.current_player)) == 0

    def get_state(self):
        """Return current board state"""
        return self.board.copy() 

File: js/checkers-game.js
class CheckersGame {
    static EMPTY = 0;
    static RED = 1;
    static BLACK = 2;
    static RED_KING = 3;
    static BLACK_KING = 4;

    constructor() {
        this.board = null;
        this.currentPlayer = null;
        this.selectedPiece = null;
        this.validMoves = [];
        this.mustCapture = false;
        this.gameHistory = [];
        this.capturedPieces = { red: [], black: [] };
        this.reset();
    }

    reset() {
        this.board = Array(8).fill().map(() => Array(8).fill(CheckersGame.EMPTY));
        this.currentPlayer = CheckersGame.RED;
        this.selectedPiece = null;
        this.validMoves = [];
        this.mustCapture = false;
        this.gameHistory = [];
        this.capturedPieces = { red: [], black: [] };
        
        // Place black pieces (top of board)
        for (let row = 0; row < 3; row++) {
            for (let col = 0; col < 8; col++) {
                if ((row + col) % 2 === 1) {
                    this.board[row][col] = CheckersGame.BLACK;
                }
            }
        }

        // Place red pieces (bottom of board)
        for (let row = 5; row < 8; row++) {
            for (let col = 0; col < 8; col++) {
                if ((row + col) % 2 === 1) {
                    this.board[row][col] = CheckersGame.RED;
                }
            }
        }
    }

    isValidPosition(row, col) {
        return row >= 0 && row < 8 && col >= 0 && col < 8;
    }

    isPiece(row, col, player) {
        if (!this.isValidPosition(row, col)) return false;
        const piece = this.board[row][col];
        return piece === player || piece === player + 2; // Regular piece or king
    }

    isOpponentPiece(row, col, player) {
        if (!this.isValidPosition(row, col)) return false;
        const opponent = player === CheckersGame.RED ? CheckersGame.BLACK : CheckersGame.RED;
        return this.isPiece(row, col, opponent);
    }

    getValidMoves(player) {
        const moves = [];
        const captures = [];

        // First, check for captures (mandatory in checkers)
        for (let row = 0; row < 8; row++) {
            for (let col = 0; col < 8; col++) {
                if (this.isPiece(row, col, player)) {
                    const pieceCaptures = this.getValidCaptures(row, col);
                    captures.push(...pieceCaptures);
                }
            }
        }

        // If captures are available, only return captures
        if (captures.length > 0) {
            this.mustCapture = true;
            return captures;
        }

        // Otherwise, get regular moves
        this.mustCapture = false;
        for (let row = 0; row < 8; row++) {
            for (let col = 0; col < 8; col++) {
                if (this.isPiece(row, col, player)) {
                    const pieceMoves = this.getValidRegularMoves(row, col);
                    moves.push(...pieceMoves);
                }
            }
        }

        return moves;
    }

    getValidCaptures(row, col) {
        const captures = [];
        const piece = this.board[row][col];
        const directions = this.getMoveDirections(piece);

        for (const [dr, dc] of directions) {
            const jumpRow = row + dr * 2;
            const jumpCol = col + dc * 2;
            const middleRow = row + dr;
            const middleCol = col + dc;

            if (this.isValidPosition(jumpRow, jumpCol) &&
                this.board[jumpRow][jumpCol] === CheckersGame.EMPTY &&
                this.isOpponentPiece(middleRow, middleCol, piece === CheckersGame.RED_KING ? CheckersGame.RED : 
                                   piece === CheckersGame.BLACK_KING ? CheckersGame.BLACK : piece)) {
                captures.push({
                    from: { row, col },
                    to: { row: jumpRow, col: jumpCol },
                    captured: { row: middleRow, col: middleCol },
                    type: 'capture'
                });
            }
        }

        return captures;
    }

    getValidRegularMoves(row, col) {
        const moves = [];
        const piece = this.board[row][col];
        const directions = this.getMoveDirections(piece);

        for (const [dr, dc] of directions) {
            const newRow = row + dr;
            const newCol = col + dc;

            if (this.isValidPosition(newRow, newCol) &&
                this.board[newRow][newCol] === CheckersGame.EMPTY) {
                moves.push({
                    from: { row, col },
                    to: { row: newRow, col: newCol },
                    type: 'move'
                });
            }
        }

        return moves;
    }

    getMoveDirections(piece) {
        const directions = [];
        
        if (piece === CheckersGame.RED) {
            directions.push([-1, -1], [-1, 1]); // Up-left, up-right
        } else if (piece === CheckersGame.BLACK) {
            directions.push([1, -1], [1, 1]); // Down-left, down-right
        } else if (piece === CheckersGame.RED_KING || piece === CheckersGame.BLACK_KING) {
            directions.push([-1, -1], [-1, 1], [1, -1], [1, 1]); // All directions
        }

        return directions;
    }

    makeMove(move) {
        const { from, to, captured, type } = move;
        const piece = this.board[from.row][from.col];
        
        // Save move to history
        this.gameHistory.push({
            move: { ...move },
            boardState: this.board.map(row => [...row]),
            currentPlayer: this.currentPlayer,
            capturedPieces: {
                red: [...this.capturedPieces.red],
                black: [...this.capturedPieces.black]
            }
        });

        // Move the piece
        this.board[from.row][from.col] = CheckersGame.EMPTY;
        this.board[to.row][to.col] = piece;

        // Handle capture
        if (type === 'capture') {
            const capturedPiece = this.board[captured.row][captured.col];
            this.board[captured.row][captured.col] = CheckersGame.EMPTY;
            
            // Add to captured pieces
            if (capturedPiece === CheckersGame.RED || capturedPiece === CheckersGame.RED_KING) {
                this.capturedPieces.red.push(capturedPiece);
            } else {
                this.capturedPieces.black.push(capturedPiece);
            }
        }

        // Handle king promotion
        if (piece === CheckersGame.RED && to.row === 0) {
            this.board[to.row][to.col] = CheckersGame.RED_KING;
        } else if (piece === CheckersGame.BLACK && to.row === 7) {
            this.board[to.row][to.col] = CheckersGame.BLACK_KING;
        }

        // Check for additional captures after a capture move
        if (type === 'capture') {
            const additionalCaptures = this.getValidCaptures(to.row, to.col);
            if (additionalCaptures.length > 0) {
                // Player must continue capturing with the same piece
                return { continueCapturing: true, piece: { row: to.row, col: to.col } };
            }
        }

        // Switch players
        this.currentPlayer = this.currentPlayer === CheckersGame.RED ? CheckersGame.BLACK : CheckersGame.RED;
        return { continueCapturing: false };
    }

    undoMove() {
        if (this.gameHistory.length === 0) return false;

        const lastState = this.gameHistory.pop();
        this.board = lastState.boardState;
        this.currentPlayer = lastState.currentPlayer;
        this.capturedPieces = lastState.capturedPieces;
        this.selectedPiece = null;
        this.validMoves = [];
        this.mustCapture = false;

        return true;
    }

    isGameOver() {
        const redMoves = this.getValidMoves(CheckersGame.RED);
        const blackMoves = this.getValidMoves(CheckersGame.BLACK);
        
        return redMoves.length === 0 || blackMoves.length === 0 ||
               this.countPieces(CheckersGame.RED) === 0 || this.countPieces(CheckersGame.BLACK) === 0;
    }

    getWinner() {
        if (!this.isGameOver()) return null;

        const redPieces = this.countPieces(CheckersGame.RED);
        const blackPieces = this.countPieces(CheckersGame.BLACK);
        const redMoves = this.getValidMoves(CheckersGame.RED);
        const blackMoves = this.getValidMoves(CheckersGame.BLACK);

        if (redPieces === 0 || redMoves.length === 0) {
            return CheckersGame.BLACK;
        } else if (blackPieces === 0 || blackMoves.length === 0) {
            return CheckersGame.RED;
        }

        return null; // Draw (shouldn't happen in checkers)
    }

    countPieces(player) {
        let count = 0;
        for (let row = 0; row < 8; row++) {
            for (let col = 0; col < 8; col++) {
                if (this.isPiece(row, col, player)) {
                    count++;
                }
            }
        }
        return count;
    }

    getBoard() {
        return this.board.map(row => [...row]);
    }

    getCurrentPlayer() {
        return this.currentPlayer;
    }

    getCapturedPieces() {
        return { ...this.capturedPieces };
    }

    getGameHistory() {
        return [...this.gameHistory];
    }
}